File: research/features/regime.py
"""
Regime Feature Generator.

Detects market regime through volatility and entropy measures:
- Parkinson Volatility: High-efficiency volatility estimator using High/Low
- Shannon Entropy: Measures randomness in price movements (noise vs trend)
"""

import numpy as np
import pandas as pd
from typing import Optional


class RegimeFeatureGenerator:
    """
    Generate regime detection features from Dollar Bar data.

    Features:
        - Parkinson Volatility: More efficient than close-to-close
        - Shannon Entropy: Low = trending, High = noisy/ranging
        - Realized Volatility: Standard close-to-close volatility
    """

    def __init__(
        self,
        vol_window: int = 24,
        entropy_window: int = 24,
        entropy_bins: int = 10,
        epsilon: float = 1e-10,
    ):
        """
        Args:
            vol_window: Rolling window for volatility calculation
            entropy_window: Rolling window for entropy calculation
            entropy_bins: Number of bins for entropy histogram
            epsilon: Small value to prevent log(0)
        """
        self.vol_window = vol_window
        self.entropy_window = entropy_window
        self.entropy_bins = entropy_bins
        self.epsilon = epsilon

    def generate(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Generate regime features.

        Args:
            df: DataFrame with at least 'high', 'low', 'close' columns

        Returns:
            DataFrame with original columns + regime features
        """
        features = df.copy()

        # 1. Parkinson Volatility
        features['parkinson_vol'] = self._parkinson_volatility(
            features['high'], features['low']
        )

        # 2. Realized Volatility (close-to-close)
        features['returns'] = features['close'].pct_change()
        features['realized_vol'] = (
            features['returns']
            .rolling(window=self.vol_window)
            .std()
        )

        # 3. Shannon Entropy
        features['shannon_entropy'] = self._rolling_entropy(features['returns'])

        # 4. Volatility Ratio (Parkinson / Realized)
        # High ratio may indicate jumps or gaps
        features['vol_ratio'] = (
            features['parkinson_vol'] /
            (features['realized_vol'] + self.epsilon)
        )

        # 5. Normalized Volatility (z-score of volatility)
        vol_mean = features['parkinson_vol'].rolling(window=self.vol_window * 4).mean()
        vol_std = features['parkinson_vol'].rolling(window=self.vol_window * 4).std()
        features['vol_zscore'] = (
            (features['parkinson_vol'] - vol_mean) /
            (vol_std + self.epsilon)
        )

        # 6. Entropy Z-score
        ent_mean = features['shannon_entropy'].rolling(window=self.entropy_window * 4).mean()
        ent_std = features['shannon_entropy'].rolling(window=self.entropy_window * 4).std()
        features['entropy_zscore'] = (
            (features['shannon_entropy'] - ent_mean) /
            (ent_std + self.epsilon)
        )

        return features

    def _parkinson_volatility(
        self,
        high: pd.Series,
        low: pd.Series,
    ) -> pd.Series:
        """
        Calculate Parkinson volatility estimator.

        More efficient than close-to-close volatility because it uses
        intrabar price range information.

        Formula: sqrt(sum(log(H/L)^2) / (4 * ln(2) * N))

        Args:
            high: High prices
            low: Low prices

        Returns:
            Rolling Parkinson volatility
        """
        # Prevent log(0) or log(negative)
        log_hl = np.log((high + self.epsilon) / (low + self.epsilon))
        log_hl_sq = log_hl ** 2

        # Rolling sum
        rolling_sum = log_hl_sq.rolling(window=self.vol_window).sum()

        # Parkinson constant: 1 / (4 * ln(2))
        parkinson_const = 1 / (4 * np.log(2))

        return np.sqrt(parkinson_const * rolling_sum / self.vol_window)

    def _rolling_entropy(self, returns: pd.Series) -> pd.Series:
        """
        Calculate rolling Shannon entropy of returns.

        Low entropy = more predictable (trending)
        High entropy = more random (noisy/ranging)

        Args:
            returns: Price returns

        Returns:
            Rolling Shannon entropy
        """
        def calc_entropy(x):
            # Remove NaN values
            x = x.dropna()
            if len(x) < 2:
                return np.nan

            # Create histogram
            hist, _ = np.histogram(x, bins=self.entropy_bins, density=True)

            # Remove zero bins and normalize
            hist = hist[hist > 0]
            if len(hist) == 0:
                return 0.0

            # Calculate entropy: -sum(p * log(p))
            # Normalize by log(bins) to get value between 0 and 1
            entropy = -np.sum(hist * np.log(hist + self.epsilon))
            max_entropy = np.log(self.entropy_bins)

            return entropy / max_entropy

        return returns.rolling(window=self.entropy_window).apply(
            calc_entropy, raw=False
        )

    def get_feature_names(self) -> list:
        """Return list of generated feature names."""
        return [
            'parkinson_vol',
            'returns',
            'realized_vol',
            'shannon_entropy',
            'vol_ratio',
            'vol_zscore',
            'entropy_zscore',
        ]

    def get_core_features(self) -> list:
        """Return core regime features for model input."""
        return [
            'parkinson_vol',
            'shannon_entropy',
            'vol_zscore',
        ]

File: etl/internal/features/regime.go
package features

import (
	"math"

	"dl-rl-btc-etl/internal/features/normalizer"
)

// RegimeFeatures contains regime detection features.
type RegimeFeatures struct {
	GarmanKlassVol float64 // Garman-Klass volatility (more efficient than Parkinson)
	RealizedVol    float64
	ShannonEntropy float64
	VolRatio       float64
	VolZScore      float64
	EntropyZScore  float64
	Skewness       float64 // 3rd moment: distribution asymmetry
	Kurtosis       float64 // 4th moment: tail thickness (excess)
}

// RegimeConfig holds configuration for regime detection.
type RegimeConfig struct {
	ParkinsonWindow int     // Window for Parkinson volatility
	EntropyWindow   int     // Window for Shannon entropy
	EntropyBins     int     // Number of bins for entropy histogram
	VolHalflife     float64 // EWM halflife for volatility z-score
	EntropyHalflife float64 // EWM halflife for entropy z-score
}

// DefaultRegimeConfig returns default regime configuration.
func DefaultRegimeConfig() RegimeConfig {
	return RegimeConfig{
		ParkinsonWindow: 24,
		EntropyWindow:   24,
		EntropyBins:     10,
		VolHalflife:     50,
		EntropyHalflife: 50,
	}
}

// Regime computes market regime features using streaming algorithms.
// It maintains state across bar updates for rolling calculations.
type Regime struct {
	config RegimeConfig

	// Garman-Klass volatility: ring buffer for GK terms
	gkBuffer *ringBuffer
	gkSum    float64

	// Realized volatility: ring buffer for returns
	returnBuffer *ringBuffer
	prevClose    float64
	prevOpen     float64

	// Shannon entropy: ring buffer for returns
	entropyBuffer *ringBuffer

	// Higher moments: Welford accumulator for skewness/kurtosis
	momentAccum *normalizer.Welford

	// EWM normalizers for z-scores
	volNormalizer     *normalizer.EWM
	entropyNormalizer *normalizer.EWM

	// Current features
	current RegimeFeatures
	count   int64
}

// ringBuffer is a simple fixed-size circular buffer.
type ringBuffer struct {
	data   []float64
	index  int
	count  int
	window int
}

func newRingBuffer(size int) *ringBuffer {
	return &ringBuffer{
		data:   make([]float64, size),
		window: size,
	}
}

func (r *ringBuffer) Push(value float64) (old float64, wasFull bool) {
	wasFull = r.count >= r.window
	if wasFull {
		old = r.data[r.index]
	}

	r.data[r.index] = value
	r.index = (r.index + 1) % r.window

	if r.count < r.window {
		r.count++
	}

	return old, wasFull
}

func (r *ringBuffer) IsFull() bool {
	return r.count >= r.window
}

func (r *ringBuffer) Values() []float64 {
	if r.count < r.window {
		return r.data[:r.count]
	}
	return r.data
}

func (r *ringBuffer) Reset() {
	r.index = 0
	r.count = 0
	for i := range r.data {
		r.data[i] = 0
	}
}

// NewRegime creates a new regime feature generator.
func NewRegime(config RegimeConfig) *Regime {
	return &Regime{
		config:            config,
		gkBuffer:          newRingBuffer(config.ParkinsonWindow),
		returnBuffer:      newRingBuffer(config.ParkinsonWindow),
		entropyBuffer:     newRingBuffer(config.EntropyWindow),
		momentAccum:       normalizer.NewWelfordWithMinSamples(30),
		volNormalizer:     normalizer.NewEWMWithHalflife(config.VolHalflife),
		entropyNormalizer: normalizer.NewEWMWithHalflife(config.EntropyHalflife),
	}
}

// Update processes a new bar and returns updated features.
func (r *Regime) Update(bar InputBar) RegimeFeatures {
	r.count++

	// === Garman-Klass Volatility ===
	// GK formula: σ² = 0.5*ln(H/L)² - (2ln2-1)*ln(C/O)²
	// More efficient estimator than Parkinson (uses all OHLC data)
	high := bar.High
	low := bar.Low
	open := bar.Open
	close := bar.Close

	// Safe division guards
	if low < Epsilon {
		low = Epsilon
	}
	if high < low {
		high = low
	}
	if open < Epsilon {
		open = Epsilon
	}
	if close < Epsilon {
		close = Epsilon
	}

	logHL := math.Log(high / low)
	logCO := math.Log(close / open)

	// GK term for this bar
	gkTerm := 0.5*logHL*logHL - (2*math.Ln2-1)*logCO*logCO

	// Update rolling sum
	old, wasFull := r.gkBuffer.Push(gkTerm)
	if wasFull {
		r.gkSum -= old
	}
	r.gkSum += gkTerm

	// GK volatility: sqrt(mean of GK terms)
	n := float64(r.gkBuffer.count)
	if n > 0 {
		// Clamp to avoid negative values from numerical errors
		meanGK := r.gkSum / n
		if meanGK < 0 {
			meanGK = 0
		}
		r.current.GarmanKlassVol = math.Sqrt(meanGK)
	}

	// === Realized Volatility (close-to-close) ===
	var ret float64
	if r.prevClose > Epsilon {
		ret = (bar.Close - r.prevClose) / r.prevClose
	}
	r.prevClose = bar.Close
	r.prevOpen = bar.Open

	// Push to return buffer for realized vol
	r.returnBuffer.Push(ret)

	// Calculate realized vol (std of returns)
	if r.returnBuffer.count >= 2 {
		returns := r.returnBuffer.Values()
		r.current.RealizedVol = stdDev(returns)
	}

	// === Vol Ratio ===
	if r.current.RealizedVol > Epsilon {
		r.current.VolRatio = r.current.GarmanKlassVol / r.current.RealizedVol
	}

	// === Shannon Entropy (Vol-Normalized) ===
	// Normalize returns by realized vol to measure distribution "shape"
	// rather than spread. This decouples entropy from volatility.
	scaledReturn := ret
	if r.current.RealizedVol > 1e-8 {
		scaledReturn = ret / r.current.RealizedVol
	}
	r.entropyBuffer.Push(scaledReturn)
	if r.entropyBuffer.IsFull() {
		r.current.ShannonEntropy = r.calcEntropy(r.entropyBuffer.Values())
	}

	// === Higher Moments (Skewness & Kurtosis) ===
	r.momentAccum.Update(ret)
	r.current.Skewness = r.momentAccum.Skewness()
	r.current.Kurtosis = r.momentAccum.Kurtosis()

	// === Z-Scores (EWM) ===
	r.current.VolZScore = r.volNormalizer.Update(r.current.GarmanKlassVol)
	r.current.EntropyZScore = r.entropyNormalizer.Update(r.current.ShannonEntropy)

	return r.current
}

// calcEntropy calculates normalized Shannon entropy of returns.
func (r *Regime) calcEntropy(returns []float64) float64 {
	if len(returns) < 2 {
		return 0
	}

	// Find min/max for binning
	minVal, maxVal := returns[0], returns[0]
	for _, v := range returns {
		if v < minVal {
			minVal = v
		}
		if v > maxVal {
			maxVal = v
		}
	}

	// Handle case where all values are the same
	rangeVal := maxVal - minVal
	if rangeVal < Epsilon {
		return 0 // All same values = zero entropy
	}

	// Create histogram
	bins := make([]int, r.config.EntropyBins)
	binWidth := rangeVal / float64(r.config.EntropyBins)

	for _, v := range returns {
		binIdx := int((v - minVal) / binWidth)
		if binIdx >= r.config.EntropyBins {
			binIdx = r.config.EntropyBins - 1
		}
		if binIdx < 0 {
			binIdx = 0
		}
		bins[binIdx]++
	}

	// Calculate entropy: -sum(p * log(p))
	n := float64(len(returns))
	var entropy float64
	for _, count := range bins {
		if count > 0 {
			p := float64(count) / n
			entropy -= p * math.Log(p)
		}
	}

	// Normalize by max entropy (log(bins))
	maxEntropy := math.Log(float64(r.config.EntropyBins))
	if maxEntropy > 0 {
		entropy /= maxEntropy
	}

	return entropy
}

// stdDev calculates sample standard deviation.
func stdDev(values []float64) float64 {
	if len(values) < 2 {
		return 0
	}

	// Calculate mean
	var sum float64
	for _, v := range values {
		sum += v
	}
	mean := sum / float64(len(values))

	// Calculate variance
	var sumSq float64
	for _, v := range values {
		diff := v - mean
		sumSq += diff * diff
	}

	variance := sumSq / float64(len(values)-1)
	return math.Sqrt(variance)
}

// CurrentFeatures returns the current feature values.
func (r *Regime) CurrentFeatures() RegimeFeatures {
	return r.current
}

// IsPrimed returns true if all normalizers are warmed up.
func (r *Regime) IsPrimed() bool {
	return r.volNormalizer.IsPrimed() && r.entropyNormalizer.IsPrimed()
}

// Reset clears all state.
func (r *Regime) Reset() {
	r.gkBuffer.Reset()
	r.returnBuffer.Reset()
	r.entropyBuffer.Reset()
	r.gkSum = 0
	r.prevClose = 0
	r.prevOpen = 0
	r.momentAccum.Reset()
	r.volNormalizer.Reset()
	r.entropyNormalizer.Reset()
	r.current = RegimeFeatures{}
	r.count = 0
}

// RegimeState represents serializable state.
type RegimeState struct {
	GKBuffer []float64
	GKSum    float64
	GKIndex  int
	GKCount  int

	ReturnBuffer []float64
	ReturnIndex  int
	ReturnCount  int
	PrevClose    float64
	PrevOpen     float64

	EntropyBuffer []float64
	EntropyIndex  int
	EntropyCount  int

	MomentState      normalizer.WelfordState
	VolNormState     normalizer.EWMState
	EntropyNormState normalizer.EWMState

	Count int64
}

// State returns serializable state snapshot.
func (r *Regime) State() RegimeState {
	return RegimeState{
		GKBuffer: append([]float64{}, r.gkBuffer.data...),
		GKSum:    r.gkSum,
		GKIndex:  r.gkBuffer.index,
		GKCount:  r.gkBuffer.count,

		ReturnBuffer: append([]float64{}, r.returnBuffer.data...),
		ReturnIndex:  r.returnBuffer.index,
		ReturnCount:  r.returnBuffer.count,
		PrevClose:    r.prevClose,
		PrevOpen:     r.prevOpen,

		EntropyBuffer: append([]float64{}, r.entropyBuffer.data...),
		EntropyIndex:  r.entropyBuffer.index,
		EntropyCount:  r.entropyBuffer.count,

		MomentState:      r.momentAccum.State(),
		VolNormState:     r.volNormalizer.State(),
		EntropyNormState: r.entropyNormalizer.State(),

		Count: r.count,
	}
}

// LoadState restores from serialized state.
func (r *Regime) LoadState(state RegimeState) {
	copy(r.gkBuffer.data, state.GKBuffer)
	r.gkSum = state.GKSum
	r.gkBuffer.index = state.GKIndex
	r.gkBuffer.count = state.GKCount

	copy(r.returnBuffer.data, state.ReturnBuffer)
	r.returnBuffer.index = state.ReturnIndex
	r.returnBuffer.count = state.ReturnCount
	r.prevClose = state.PrevClose
	r.prevOpen = state.PrevOpen

	copy(r.entropyBuffer.data, state.EntropyBuffer)
	r.entropyBuffer.index = state.EntropyIndex
	r.entropyBuffer.count = state.EntropyCount

	r.momentAccum.LoadState(state.MomentState)
	r.volNormalizer.LoadState(state.VolNormState)
	r.entropyNormalizer.LoadState(state.EntropyNormState)

	r.count = state.Count
}
